File: xkcd/src/main/java/xyz/jienan/xkcd/model/XkcdModel.java
package xyz.jienan.xkcd.model;


import java.util.List;

import io.reactivex.Observable;
import io.reactivex.Single;
import io.reactivex.schedulers.Schedulers;
import io.reactivex.subjects.PublishSubject;
import okhttp3.ResponseBody;
import xyz.jienan.xkcd.base.network.NetworkService;
import xyz.jienan.xkcd.base.network.XkcdAPI;
import xyz.jienan.xkcd.model.persist.BoxManager;
import xyz.jienan.xkcd.model.util.XkcdExplainUtil;

import static xyz.jienan.xkcd.base.network.NetworkService.XKCD_BROWSE_LIST;
import static xyz.jienan.xkcd.base.network.NetworkService.XKCD_EXPLAIN_URL;
import static xyz.jienan.xkcd.base.network.NetworkService.XKCD_SEARCH_SUGGESTION;
import static xyz.jienan.xkcd.base.network.NetworkService.XKCD_TOP;
import static xyz.jienan.xkcd.base.network.NetworkService.XKCD_TOP_SORT_BY_THUMB_UP;

public class XkcdModel {

    private static final int SLICE = 400;
    private static XkcdModel xkcdModel;
    private final BoxManager boxManager = BoxManager.getInstance();
    private final PublishSubject<XkcdPic> picsPipeline = PublishSubject.create();
    private final XkcdAPI xkcdApi = NetworkService.getXkcdAPI();

    private XkcdModel() {
        // no public constructor
    }

    public static XkcdModel getInstance() {
        if (xkcdModel == null) {
            xkcdModel = new XkcdModel();
        }
        return xkcdModel;
    }

    public void push(XkcdPic pic) {
        picsPipeline.onNext(pic);
    }

    public Observable<XkcdPic> observe() {
        return picsPipeline;
    }

    public Observable<XkcdPic> loadLatest() {
        return xkcdApi.getLatest()
                .subscribeOn(Schedulers.io())
                .map(boxManager::updateAndSave);
    }

    public Observable<XkcdPic> loadXkcd(long index) {
        return xkcdApi.getXkcdList(XKCD_BROWSE_LIST, (int) index, 0, 1)
                .subscribeOn(Schedulers.io())
                .flatMap(xkcdPics -> {
                    if (xkcdPics == null || xkcdPics.isEmpty()) {
                        return xkcdApi.getComics(index).subscribeOn(Schedulers.io());
                    } else {
                        return Observable.just(xkcdPics.get(0));
                    }
                }).map(boxManager::updateAndSave);
    }

    /**
     * fast loading all xkcd pics
     *
     * @param latestIndex
     * @return
     */
    public Observable<Boolean> fastLoad(int latestIndex) {
        return Observable.range(0, (latestIndex - 1) / SLICE + 1)
                .subscribeOn(Schedulers.io())
                .map(i -> i * 400 + 1)
                .flatMap(startIndex -> Observable.just(boxManager.getValidXkcdInRange(startIndex, startIndex + SLICE - 1))
                        .filter(xkcdPics -> xkcdPics.size() != SLICE)
                        .filter(xkcdPics -> {
                            if (xkcdPics.isEmpty()) {
                                return true;
                            } else {
                                final int size = xkcdPics.size();
                                if (startIndex <= 404 && startIndex + SLICE > 404 && size != SLICE - 1) {
                                    return true;
                                } else if (startIndex > latestIndex - SLICE + 1 && startIndex + size - 1 != latestIndex) {
                                    return true;
                                }
                            }
                            return false;
                        }).flatMap(ignored -> loadRange(startIndex, SLICE)))
                .map(ignored -> Boolean.TRUE);
    }

    public Observable<List<XkcdPic>> getThumbUpList() {
        return xkcdApi.getTopXkcds(XKCD_TOP, XKCD_TOP_SORT_BY_THUMB_UP)
                .subscribeOn(Schedulers.io())
                .map(boxManager::updateAndSave);
    }

    /**
     * @param start
     * @param range
     * @return last index.
     */
    public Observable<List<XkcdPic>> loadRange(long start, long range) {
        return xkcdApi.getXkcdList(XKCD_BROWSE_LIST, (int) start, 0, (int) range)
                .subscribeOn(Schedulers.io())
                .map(boxManager::updateAndSave);
    }

    /**
     * @param index
     * @return thumb up count
     */
    public Observable<Long> thumbsUp(long index) {
        return xkcdApi.thumbsUp(NetworkService.XKCD_THUMBS_UP, (int) index)
                .subscribeOn(Schedulers.io())
                .doOnSubscribe(ignored -> boxManager.likeXkcd(index))
                .map(boxManager::updateAndSave)
                .map(xkcdPic -> xkcdPic.thumbCount);
    }

    public Single<Boolean> validateXkcdList(List<XkcdPic> xkcdList) {
        return Observable.fromIterable(xkcdList)
                .subscribeOn(Schedulers.io())
                .filter(xkcdPic -> xkcdPic == null || xkcdPic.width == 0 || xkcdPic.height == 0)
                .toSortedList()
                .flatMap(xkcdPics -> {
                    if (xkcdPics.isEmpty()) {
                        return Single.just(0);
                    } else {
                        return loadRange(xkcdPics.get(0).num,
                                xkcdPics.get(xkcdPics.size() - 1).num).singleOrError();
                    }
                }).map(ignore -> Boolean.TRUE);
    }

    public List<XkcdPic> getFavXkcd() {
        return boxManager.getFavXkcd();
    }

    public XkcdPic loadXkcdFromDB(long index) {
        return boxManager.getXkcd(index);
    }

    public List<XkcdPic> loadXkcdFromDB(long start, long end) {
        return boxManager.getXkcdInRange(start, end);
    }

    public XkcdPic updateSize(long index, int width, int height) {
        XkcdPic xkcdPic = boxManager.getXkcd(index);
        if (xkcdPic != null && width > 0 && height > 0) {
            xkcdPic.width = width;
            xkcdPic.height = height;
            boxManager.saveXkcd(xkcdPic);
        }
        return xkcdPic;
    }

    public Observable<XkcdPic> fav(long index, boolean isFav) {
        final XkcdPic xkcdPicInBox = boxManager.favXkcd(index, isFav);
        if (xkcdPicInBox.width == 0 || xkcdPicInBox.height == 0) {
            return loadXkcd(index);
        } else {
            return Observable.just(xkcdPicInBox);
        }
    }

    public Observable<List<XkcdPic>> search(String query) {
        return xkcdApi.getXkcdsSearchResult(XKCD_SEARCH_SUGGESTION, query)
                .subscribeOn(Schedulers.io())
                .map(boxManager::updateAndSave);
    }

    public Observable<String> loadExplain(long index, long latestIndex) {
        final String url = XKCD_EXPLAIN_URL + index;
        final Observable<ResponseBody> observable = latestIndex - index < 10 && latestIndex - index >= 0 ?
                xkcdApi.getExplainWithShortCache(url, 1800 * (latestIndex - index + 1))
                : xkcdApi.getExplain(url);

        return observable.subscribeOn(Schedulers.io())
                .map(responseBody -> XkcdExplainUtil.getExplainFromHtml(responseBody, url));
    }
}

File: xkcd/src/main/java/xyz/jienan/xkcd/model/XkcdModel.kt
package xyz.jienan.xkcd.model


import io.reactivex.Maybe
import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.schedulers.Schedulers
import io.reactivex.subjects.PublishSubject
import timber.log.Timber
import xyz.jienan.xkcd.base.network.NetworkService
import xyz.jienan.xkcd.base.network.XKCD_BROWSE_LIST
import xyz.jienan.xkcd.base.network.XKCD_EXPLAIN_URL
import xyz.jienan.xkcd.base.network.XKCD_TOP_SORT_BY_THUMB_UP
import xyz.jienan.xkcd.model.persist.BoxManager
import xyz.jienan.xkcd.model.util.XkcdExplainUtil
import java.util.concurrent.TimeUnit

object XkcdModel {

    var localizedUrl: String = ""

    private const val SLICE = 400

    private val picsPipeline = PublishSubject.create<XkcdPic>()

    val thumbUpList: Observable<List<XkcdPic>>
        get() = NetworkService.xkcdAPI
                .getTopXkcds(XKCD_TOP_SORT_BY_THUMB_UP)
                .subscribeOn(Schedulers.io())
                .map { BoxManager.updateAndSave(it) }

    val favXkcd: List<XkcdPic>
        get() = BoxManager.favXkcd

    val untouchedList: List<XkcdPic>
        get() = BoxManager.untouchedComicsList

    fun push(pic: XkcdPic) {
        picsPipeline.onNext(pic)
    }

    fun observe(): Observable<XkcdPic> = picsPipeline

    fun loadLatest(): Observable<XkcdPic> = NetworkService.xkcdAPI
            .latest
            .subscribeOn(Schedulers.io())
            .map { BoxManager.updateAndSave(it) }

    fun loadXkcd(index: Long): Observable<XkcdPic> = NetworkService.xkcdAPI
            .getXkcdList(XKCD_BROWSE_LIST, index.toInt(), 0, 1)
            .subscribeOn(Schedulers.io())
            .flatMap { xkcdPics ->
                if (xkcdPics.isEmpty()) {
                    NetworkService.xkcdAPI.getComics(index).subscribeOn(Schedulers.io())
                } else {
                    Observable.just<XkcdPic>(xkcdPics[0])
                }
            }.map { BoxManager.updateAndSave(it) }

    /**
     * fast loading all xkcd pics
     *
     * @param latestIndex
     * @return
     */
    fun fastLoad(latestIndex: Int): Single<Long> {

        val sectionCount = (latestIndex - 1) / SLICE + 1

        val lastSectionStartIndex = SLICE * (sectionCount - 1) + 1

        return Observable.range(0, (latestIndex - 1) / SLICE + 1)
                .subscribeOn(Schedulers.io())
                .map { i -> i * 400 + 1 }
                .flatMap { startIndex ->
                    Observable.just(BoxManager.getValidXkcdInRange(startIndex.toLong(), startIndex + SLICE - 1L))
                            .flatMap {
                                Timber.d("Start Index $startIndex, list size :${it.size}")
                                if (startIndex == 401 && it.size == 399 || it.size == SLICE) {
                                    Observable.just(true)
                                } else if (lastSectionStartIndex == startIndex && it.size == latestIndex - startIndex + 1) {
                                    Observable.just(true)
                                } else {
                                    loadRange(startIndex.toLong(), SLICE.toLong()).map { true }
                                }
                            }
                }
                .doOnNext { Timber.d("XkcdFastLoad") }
                .take(sectionCount.toLong())
                .count()
    }

    /**
     * @param start
     * @param range
     * @return last index.
     */
    fun loadRange(start: Long, range: Long, reversed: Int = 0): Observable<List<XkcdPic>> = NetworkService.xkcdAPI
            .getXkcdList(XKCD_BROWSE_LIST, start.toInt(), reversed, range.toInt())
            .subscribeOn(Schedulers.io())
            .map { BoxManager.updateAndSave(it) }

    /**
     * @param index
     * @return thumb up count
     */
    fun thumbsUp(index: Long): Observable<Long> = NetworkService.xkcdAPI
            .thumbsUp(index.toInt())
            .subscribeOn(Schedulers.io())
            .doOnSubscribe { BoxManager.likeXkcd(index) }
            .map { BoxManager.updateAndSave(it) }
            .map { it.thumbCount }

    fun validateXkcdList(xkcdList: List<XkcdPic>): Single<Boolean> = Observable.fromIterable(xkcdList)
            .subscribeOn(Schedulers.io())
            .filter { xkcdPic -> xkcdPic.width == 0 || xkcdPic.height == 0 }
            .toSortedList()
            .flatMap { xkcdPics ->
                if (xkcdPics.isEmpty()) {
                    Single.just(0)
                } else {
                    loadRange(xkcdPics[0].num, xkcdPics[xkcdPics.size - 1].num)
                            .singleOrError()
                }
            }.map { true }

    fun loadXkcdFromDB(index: Long) = BoxManager.getXkcd(index)

    fun loadXkcdFromDB(start: Long, end: Long) = BoxManager.getXkcdInRange(start, end)

    fun updateSize(index: Long, width: Int, height: Int) =
            BoxManager.getXkcd(index).also {
                if (it != null && width > 0 && height > 0) {
                    it.width = width
                    it.height = height
                    BoxManager.saveXkcd(it)
                }
            }

    fun fav(index: Long, isFav: Boolean): Observable<XkcdPic> {
        val xkcdPicInBox = BoxManager.favXkcd(index, isFav)
        return if (xkcdPicInBox == null || xkcdPicInBox.width == 0 || xkcdPicInBox.height == 0) {
            loadXkcd(index)
        } else {
            Observable.just(xkcdPicInBox)
        }
    }

    fun search(query: String): Observable<List<XkcdPic>> =
            NetworkService.xkcdAPI
                    .getXkcdsSearchResult(query)
                    .subscribeOn(Schedulers.io())
                    .map { BoxManager.updateAndSave(it) }

    fun loadExplain(index: Long, latestIndex: Long): Observable<String> {
        val url = XKCD_EXPLAIN_URL + index
        val observable = if (latestIndex - index in 0..9)
            NetworkService.xkcdAPI
                    .getExplainWithShortCache(url, 1800 * (latestIndex - index + 1))
        else
            NetworkService.xkcdAPI.getExplain(url)

        return observable.subscribeOn(Schedulers.io())
                .map { XkcdExplainUtil.getExplainFromHtml(it, url) }
    }

    fun loadLocalizedXkcd(index: Long): Maybe<XkcdPic> {
        return if (localizedUrl.isBlank()) {
            Maybe.error(Exception("Local xkcd not configured"))
        } else {
            NetworkService.xkcdAPI
                    .getLocalizedXkcd(localizedUrl.format(index))
                    .timeout(3000L, TimeUnit.MILLISECONDS)
                    .toMaybe()
        }
    }
}